components/trace_metric/src/metric.rs

```diff
@@ -17,18 +23,48 @@ pub enum Metric {
 
 impl Metric {
     #[inline]
-    pub fn number(name: String, val: usize) -> Self {
-        Metric::Number(MetricValue { name, val })
+    pub fn number(name: String, val: usize, op: Option<MetricOp>) -> Self {
+        Metric::Number(MetricValue { name, val, op })
     }
 
     #[inline]
-    pub fn duration(name: String, val: Duration) -> Self {
-        Metric::Duration(MetricValue { name, val })
+    pub fn duration(name: String, val: Duration, op: Option<MetricOp>) -> Self {
+        Metric::Duration(MetricValue { name, val, op })
     }
 
     #[inline]
-    pub fn boolean(name: String, val: bool) -> Self {
-        Metric::Boolean(MetricValue { name, val })
+    pub fn boolean(name: String, val: bool, op: Option<MetricOp>) -> Self {
+        Metric::Boolean(MetricValue { name, val, op })
+    }
+
+    #[inline]
+    pub fn name(&self) -> &str {
+        match self {
+            Self::Boolean(v) => &v.name,
+            Self::Number(v) => &v.name,
+            Self::Duration(v) => &v.name,
+        }
+    }
+
+    #[inline]
+    pub fn op(&self) -> &Option<MetricOp> {
+        match self {
+            Self::Boolean(v) => &v.op,
+            Self::Number(v) => &v.op,
+            Self::Duration(v) => &v.op,
+        }
+    }
+
+    // Add performs value add when metrics are same type,
+    // If their types are different, do nothing.
+    #[inline]
+    pub fn add(&mut self, rhs: &Self) {
+        match (self, rhs) {
+            (Self::Boolean(lhs), Self::Boolean(rhs)) => lhs.val |= rhs.val,
+            (Self::Number(lhs), Self::Number(rhs)) => lhs.val += rhs.val,
+            (Self::Duration(lhs), Self::Duration(rhs)) => lhs.val += rhs.val,
+            _ => {}
+        }
```

Comment:
Shall we panic when adding two Metric with different types? Otherwise, a new branch for a new type maybe be omitted here.

Reply:
OK.